## Cross-family review (xAI / Grok) of PR #6884

**Head:** `2fa7366f07007181559e440fc68b9b9097706754` (`claude/infra-6830-venv-debt`)
**Author family:** Anthropic (claude lane)
**Reviewer family:** xAI (grok-4.6, native grok CLI)
**Issue:** #6830 venv-side asks (manifest debt, recovery runbook, venv-importability canary). Ask #2 (background-kill attribution) is correctly scoped out on the issue.

Adversarial checks first. I re-ran an independent import scan, executed the canary against the live primary venv, walked the runbook, inspected `delegate.py` +56, ran the touched tests, and spot-checked the mutation claim.

---

### 1. Requirements changes vs reality

Independent AST scan of `scripts/` + `.mcp/` against `requirements.txt`.

**Additions are used** (not invented comments):

| Distribution | Floor in this PR | Lock (`requirements-lock.txt`) | Import sites found |
|---|---|---|---|
| `psutil` | `>=7.2` | `psutil==7.2.2` | `livenessprobe.py`, `watchdog.py`, `agent_monitor_router.py`, `thread_handoff.py`, `topology.py` |
| `filelock` | `>=3.24` | `filelock==3.29.4` | `pipeline/state.py`, `stress_annotator.py`, `orchestration/curriculum_coordinator.py` |
| `rapidfuzz` | `>=3.14` | `RapidFuzz==3.14.5` | `.mcp/servers/sources/server.py`, `build/linear_pipeline.py` |
| `lxml` | `>=6.1` | `lxml==6.1.1` | `ingest/esum_abbyy_parser.py` |
| `pypdf` | `>=6.15` | `pypdf==6.15.0` | `open_model_data/phase3_*.py`, `rag/scrape_diasporiana.py`, `rag/extract_text.py` |
| `PyMuPDF` | `>=1.28` | `PyMuPDF==1.28.0` | `import pymupdf` in `extract_images.py`, `annotate_images.py`, `poc_pair_page.py`, `analyze_textbook_pages.py`, `images_router.py`, `extract_text.py` |
| `ruamel.yaml` | `>=0.19` | `ruamel.yaml==0.19.1` | `fix_plan_latin_refs.py`, `convert_seminar_plans_v6.py` |

Diff vs `origin/main` is add-only — nothing previously declared was dropped. Floors are lock major.minor, matching the rest of `requirements.txt`. Remaining undeclared third-party names from the scan are first-party path identities or already-transitive/optional (numpy via torch, pydantic/starlette via fastapi, `pycookiecheat` already documented as a known-harmless lock exception). The PR's "collection-breaking + issue-named" scope is honest.

---

### 2. `check_venv_integrity.py` — read-only? catches the live #6818/#6830 pytest case?

**Shebang check: yes, on the live primary venv.** I ran the checker with an isolated `--state-dir` (no primary telemetry write):

```
cbor2   -> .../learn-ukrainian-infra-private/.worktrees/dispatch/codex/465-landing-attestation-fix/.venv/bin/python3  exists=false
py.test -> (same deleted worktree path)  exists=false
pytest  -> (same deleted worktree path)  exists=false
```

Those three are regular files using pip's `#!/bin/sh` + `'''exec' '…python3' "$0" "$@"` form. `_SHEBANG_EXEC_RE` extracts the embedded path; `os.path.realpath` of a nonexistent path is pure normalization and does not equal `realpath(.venv/bin/python)` (`~/.pyenv/versions/3.12.8/bin/python3.12`). 31 other launchers were healthy; 0 `/usr/bin/env` shebangs; `ruff` is a native binary and is correctly skipped. The deleted worktree path does not exist on disk (`exists: false`). This is the documented live case, not a synthetic one.

**Read-only against the venv it checks: yes, with one footnote.** The module itself has no `write`/`unlink`/`chmod`/`mkdir` against `venv_dir`. It only:

- reads `bin/` launcher text
- runs `[venv_python, "-c", import-probe]`
- calls sibling `_append_event` into `state_dir` (telemetry JSONL, default `data/telemetry/venv-integrity/`, not the venv)

`/api/health` does **not** include this canary (only `/api/orient` health, TTL 15s). Fail-open on probe exception matches the siblings. Repair command is printed, never executed.

Footnote (non-blocking): the sentinel subprocess can write `__pycache__/*.pyc` into the venv. `PYTHONDONTWRITEBYTECODE=1` on that `subprocess.run` would make the "never mutate the venv" claim strict on the dispatch/API path.

---

### 3. Recovery runbook — runnable in order on a clean host?

The broken `pip install -e .` step is correctly removed. `pyproject.toml` has no `[build-system]` / `[tool.setuptools]` table (verified). The replacement (`uv venv --python 3.12.8` + `uv pip install -r requirements.txt -r requirements-dev.txt`) plus the `.pth` explanation matches the issue comments.

Walk notes:

- §1 first fence assumes **pyenv** (pre-existing) **and `uv`** (new). `CONTRIBUTING.md` / `README.md` still document `python -m venv` + `pip install -r requirements.txt` and do **not** list `uv` as a prerequisite. On a host that has only the documented prereqs, `uv venv` fails. On this machine, `uv python find 3.12.8` from `/tmp` correctly resolves to the pyenv 3.12.8 with sqlite extensions — so the sequence works *here*.
- The `.pth` (required by the issue for non-pytest `scripts.*` / bare `audit.*` imports) is documented after the first fence, not inside it. A copy-paste of only the fenced block leaves project imports broken outside pytest.
- `requirements-dev.txt` already includes `-r requirements.txt`, so the dual `-r` is redundant (harmless). `.venv/bin/python -m pip install --upgrade pip` is leftover from the old sequence; `uv pip` does not use that pip.

The runbook is a real improvement over a command that fails outright. The two walkability gaps are fixable follow-ups, not a revert of the content.

---

### 4. `scripts/delegate.py` +56 — what it does; in-scope?

Two call sites, copied from `check_node_modules_integrity`:

1. `_warn_venv_integrity()` from `cmd_dispatch` — advisory print, **never blocks** a dispatch (unlike `_resolve_primary_integrity_error`).
2. Post-worker sweep in `_run_worker` — on alert, appends `venv_integrity_post_worker` to dispatch events with `task_id`/`agent` still known.

Plus the autouse stub in `tests/test_delegate.py` so a live-broken primary venv does not pollute hermetic event-file assertions.

This is issue ask #3 ("health-canary alongside the existing primary-integrity canary") implemented via the established three-site sibling pattern (delegate pre-dispatch + post-worker + API orient). **Not scope creep.** Ask #2 stays on the issue.

---

### 5. Tests + mutation

Touched files, `PYTEST_ADDOPTS="-p no:cacheprovider"`, primary interpreter:

```
tests/test_check_venv_integrity.py ................   16 passed
tests/test_delegate.py             .................. 242 passed
258 passed in 19.89s
```

Also: `tests/test_orient_api.py::test_health_includes_core_bare_canary` and `test_health_canaries_do_not_apply_general_repairs` — 2 passed. `ruff check` clean on every touched Python file.

**Mutation spot-check** (claim: reverting a detection path kills 2–6 tests). I short-circuited `if resolved != venv_python` to `if False and …` in `find_broken_console_scripts`, then restored via `git checkout`. Result: **6 failed, 10 passed**:

- `test_launcher_pointing_at_deleted_path_is_detected`
- `test_launcher_pointing_at_a_different_existing_interpreter_is_detected`
- `test_broken_launcher_is_alerted_and_recorded`
- `test_py_test_and_pytest_dedupe_to_one_package`
- `test_alert_event_names_running_dispatches`
- `test_cli_exit_codes`

After restore: 16 passed again. Claim holds. Working tree left clean.

Pytest fastlane on this PR is green (it only runs changed tests). The full shard is not.

---

### Findings

1. **[bug, blocking] API import-pinning gate is red, caused by this PR.** `_venv_integrity_canary` lazy-imports `scripts.audit.check_venv_integrity` and `audit.check_venv_integrity` (`scripts/api/main.py:1062` / `:1064`) but does not register either name in `scripts/api/preload.py` `PRELOAD_MODULES`. The siblings this canary copies **are** registered (`check_primary_integrity` / `check_node_modules_integrity` at preload.py:44–47).

   Reproduced locally and in CI:

   ```
   FAILED tests/api/test_import_pinning.py::test_ast_imports_and_loaders
   AssertionError: Found 2 import pinning violation(s).
     - Uncovered nested import 'scripts.audit.check_venv_integrity' in scripts/api/main.py:1062
     - Uncovered nested import 'audit.check_venv_integrity' in scripts/api/main.py:1064
   ```

   CI: Python (pytest) [3/4] `1 failed, 4272 passed, 18 skipped` → CI Gate red. Fastlane missed it because that test is not in the changed-file set.

   **Fix:** add both module names to `PRELOAD_MODULES` next to the two sibling pairs (preferred — matches the existing canaries). `# lazy-ok:` would also satisfy the AST guard but would leave the canary out of the API preload closure.

2. **[suggestion] Recovery §1 now requires `uv`, which is not a documented host prerequisite.** `CONTRIBUTING.md` still uses `python -m venv` + `pip install -r requirements.txt`. Either add an explicit `uv` install step, or keep the portable venv+pip sequence now that the manifest is complete. Also consider putting the `.pth` (`"$PWD"` + `"$PWD/scripts"`) inside the first copy-paste block.

3. **[nit] Sentinel set does not include `lxml` / `pypdf` / `pymupdf` / `ruamel`.** Empty-venv is still caught via `yaml`. After this PR those four are in `requirements.txt`, so the collection-break class is closed at install time. Optional follow-up if the canary should also detect a partial restore that only missed the document parsers.

4. **[nit] Sentinel subprocess can write `.pyc` into the primary venv.** Set `PYTHONDONTWRITEBYTECODE=1` on the probe env if "detection only / never mutate the venv" is meant strictly on the dispatch path.

---

**VERDICT: CHANGES REQUESTED**

The venv-side work is otherwise sound: manifest additions are real and lock-aligned, the shebang canary catches the live deleted-worktree `pytest`/`py.test`/`cbor2` case with zero false positives on the other 31 launchers, the runbook no longer recommends a command that cannot succeed, and `delegate.py` +56 is in-scope sibling wiring. Do not merge until finding 1 is fixed and `tests/api/test_import_pinning.py::test_ast_imports_and_loaders` is green on the same head.

--- a/scripts/api/preload.py
+++ b/scripts/api/preload.py
@@ -45,6 +45,8 @@ PRELOAD_MODULES = [
     "audit.check_primary_integrity",
     "scripts.audit.check_node_modules_integrity",
     "audit.check_node_modules_integrity",
+    "scripts.audit.check_venv_integrity",
+    "audit.check_venv_integrity",
     "scripts.build.phase_constants",
     "agent_runtime.adapters.gemini",
     "research_quality",

